plugins/hypervisors/kvm/src/main/java/org/apache/cloudstack/kvm/ha/KvmHaAgentClient.java

```diff
@@ -0,0 +1,295 @@
+     */
+    protected HttpResponse retryHttpRequest(String url, HttpRequestBase httpReq, HttpClient client) {
+        LOGGER.warn(String.format("Failed to execute HTTP %s request [URL: %s]. Executing the request again.", httpReq.getMethod(), url));
+        HttpResponse response = retryUntilGetsHttpResponse(url, httpReq, client);
+
+        if (response == null) {
+            LOGGER.error(String.format("Failed to execute HTTP %s request [URL: %s].", httpReq.getMethod(), url));
+            return response;
+        }
+
+        int statusCode = response.getStatusLine().getStatusCode();
+        if (statusCode < HttpStatus.SC_OK || statusCode >= HttpStatus.SC_MULTIPLE_CHOICES) {
+            LOGGER.error(
+                    String.format("Failed to get VMs information with a %s request to URL '%s'. The expected HTTP status code is '%s' but it got '%s'.", HttpGet.METHOD_NAME, url,
+                            EXPECTED_HTTP_STATUS, statusCode));
+            return null;
+        }
+
+        LOGGER.debug(String.format("Successfully executed HTTP %s request [URL: %s].", httpReq.getMethod(), url));
+        return response;
+    }
+
+    protected HttpResponse retryUntilGetsHttpResponse(String url, HttpRequestBase httpReq, HttpClient client) {
+        for (int attempt = 1; attempt < MAX_REQUEST_RETRIES + 1; attempt++) {
+            try {
+                TimeUnit.SECONDS.sleep(WAIT_FOR_REQUEST_RETRY);
+                LOGGER.debug(String.format("Retry HTTP %s request [URL: %s], attempt %d/%d.", httpReq.getMethod(), url, attempt, MAX_REQUEST_RETRIES));
+                return client.execute(httpReq);
+            } catch (IOException | InterruptedException e) {
+                String errorMessage = String.format("Failed to execute HTTP %s request retry attempt %d/%d [URL: %s] due to exception %s",
+                        httpReq.getMethod(), attempt, MAX_REQUEST_RETRIES, url, e);
+                LOGGER.error(errorMessage);
+            }
+        }
+        return null;
+    }
+
+    /**
+     * Processes the response of request GET System ID as a JSON object.<br>
+     * Json example: {"count": 3, "virtualmachines": ["r-123-VM", "v-134-VM", "s-111-VM"]}<br><br>
+     *
+     * Note: this method can return NULL JsonObject in case HttpResponse is NULL.
+     */
+    protected JsonObject processHttpResponseIntoJson(HttpResponse response) {
+        InputStream in;
+        String jsonString;
+        if (response == null) {
+            return null;
+        }
+        try {
+            in = response.getEntity().getContent();
+            BufferedReader streamReader = new BufferedReader(new InputStreamReader(in, StandardCharsets.UTF_8));
+            jsonString = streamReader.readLine();
+        } catch (UnsupportedOperationException | IOException e) {
+            throw new CloudRuntimeException("Failed to process response", e);
+        }
+
+        return new JsonParser().parse(jsonString).getAsJsonObject();
```

Comment:
```suggestion
        if (response == null) {
            return null;
        }

        try {
            InputStream in = response.getEntity().getContent();
            BufferedReader streamReader = new BufferedReader(new InputStreamReader(in, StandardCharsets.UTF_8));
            String jsonString = streamReader.readLine();
            return new JsonParser().parse(jsonString).getAsJsonObject();
        } catch (UnsupportedOperationException | IOException e) {
            throw new CloudRuntimeException("Failed to process response", e);
        }
```

Reply:
If needed, could we improve the exception with some `response` context?

Reply:
Seems to me that this method will be called several times, should we extract this `new JsonParser()` to a constant?

Reply:
Good point, I am going to update it on the next commit :+1:

Reply:
@GutoVeronezi do you mean something like this, right?
```
public class KvmHaAgentClient {
    private static final JsonParser JSON_PARSER = new JsonParser();
    ...
    ...
    ...
    @Nullable
    protected JsonObject processHttpResponseIntoJson(HttpResponse response) {
        if (response == null) {
            return null;
        }
        try {
            InputStream in = response.getEntity().getContent();
            BufferedReader streamReader = new BufferedReader(new InputStreamReader(in, StandardCharsets.UTF_8));
            return JSON_PARSER.parse(streamReader.readLine()).getAsJsonObject();
        } catch (UnsupportedOperationException | IOException e) {
            throw new CloudRuntimeException("Failed to process response", e);
        }
    }
    ...
    ...
}
```